Add a JavaScript tooling and best practices guide

robertknight opened the pull request:
Add a guide to the use of JavaScript in our frontend projects. This is
intended to provide an overview of the tools and libraries we use, as
well as our conventions in different areas (eg. documentation, testing).

robertknight committed:
Add a JavaScript tooling and best practices guide

Add a guide to the use of JavaScript in our frontend projects. This is
intended to provide an overview of the tools and libraries we use, as
well as our conventions in different areas (eg. documentation, testing).

lyzadanger commented on docs/js-guide.md:
```diff
@@ -0,0 +1,143 @@
+
+_The native features typically enable better tooling._
+
+Projects may use any ES language features which meet the following criteria:
```
ALL of the following?

robertknight replied:
Indeed. Will update.

lyzadanger commented on docs/js-guide.md:
```diff
@@ -0,0 +1,143 @@
+
+## Formatting code
+
+All of our projects use [Prettier](https://prettier.io) to format JavaScript code,
```
I wonder about the maintainability of this chunk of documentation. Will we remember to update it if we change the way we configure prettier? Would a link to an example `prettierrc` in one of our repos do?

robertknight replied:
I think we'll be OK. The only scenario where I envisage deviating additionally from Prettier's defaults is if they make a major change in a new release and we decide that we don't want to roll with it. Otherwise we try to keep the config as minimal as possible.

lyzadanger commented on docs/js-guide.md:
```diff
@@ -0,0 +1,143 @@
+
+### prop-types
+
+The [prop-types](https://reactjs.org/docs/typechecking-with-proptypes.html) package
```
This section jumps into the assumption that we're talking about preact components. Perhaps be a little more explicit about that?

robertknight replied:
Ah, yes. Good point.

lyzadanger reviewed (commented):
I think this is generally looking fine. I'd lean more toward "let's publish something quickly and evolve it"  versus "it's gotta be the end-all-be-all document," and it certainly meets my criteria at present...

robertknight committed:
Update following PR feedback

 - Move the section on prop-types to the "UI components" section to make
   it clear that it refers to Preact components

 - Make it clear that _all_ rather than just _some_ of the criteria need
   to be met before we can use a language feature

robertknight commented:
I've addressed the comments you added and based on this:

> I'd lean more toward "let's publish something quickly and evolve it" versus "it's gotta be the end-all-be-all document," and it certainly meets my criteria at present...

I've marked this PR as ready for review so we can land an initial version and add new things as we go.

lyzadanger reviewed (approved):
Ship it and we'll keep making it better over time!
